rust/stracciatella/src/vfs/android.rs

```diff
@@ -75,27 +93,42 @@ impl AssetManagerFs {
             if want == "." || want == ".." {
                 return Err(io::Error::new(
                     io::ErrorKind::InvalidInput,
-                    "special path components are not supported",
+                    "AssetManagerFs: Special path components are not supported",
                 ));
             }
             for candidate in candidates {
                 let candidate_cstring = Self::path_to_cstring(&candidate)?;
-                let dir_contents = crate::android::list_asset_dir(&candidate).map_err(|e| {
-                    io::Error::new(
-                        io::ErrorKind::Other,
-                        format!("AssetManagerFs: JNI Error: `{:?}`", e),
-                    )
-                })?;
-                if self.asset_manager.open_dir(&candidate_cstring).is_some() {
-                    for entry in dir_contents {
-                        let is_match = entry
-                            .file_name()
-                            .and_then(|x| x.to_str())
-                            .map(|x| want == Nfc::caseless(x).as_str())
-                            .unwrap_or(false);
-                        if is_match {
-                            next.push(candidate.join(&entry));
-                        }
+                let entries = if let Some(cache_entry) = canonicalization_cache.get(&candidate) {
+                    cache_entry
+                } else if self.asset_manager.open_dir(&candidate_cstring).is_none() {
+                    canonicalization_cache.put(candidate.clone(), vec![]);
+                    canonicalization_cache
+                        .get(&candidate)
+                        .expect("we should be able to get a cache key that was just set")
+                } else {
+                    let entries: Vec<_> = crate::android::list_asset_dir(&candidate)
+                        .map_err(|e| {
+                            io::Error::new(
+                                io::ErrorKind::Other,
+                                format!("AssetManagerFs: JNI Error: `{:?}`", e),
+                            )
+                        })?
+                        .iter()
+                        .flat_map(|path_buf| {
+                            path_buf
+                                .to_str()
+                                .map(|s| (Nfc::caseless(s), path_buf.into()))
+                        })
+                        .collect();
+                    canonicalization_cache.put(candidate.clone(), entries);
+                    canonicalization_cache
+                        .get(&candidate)
+                        .expect("we should be able to get a cache key that was just set")
```

Comment:
I don't understand this part. Is this testing code, or what use-cases does it cover?

Reply:
It's needed because we move the ownership into `canonicalization_cache`, so we cannot access the `entries` variable anymore. This is why it is necessary to use `.get` to get a reference to it.

Reply:
I get it now. I am not used to the implicit `return`